esp/services/ldapenvironment/ldapenvironmentService.cpp

```diff
@@ -0,0 +1,204 @@
+            secmgr = dynamic_cast<CLdapSecManager*>(context.querySecManager());
+            if (secmgr == nullptr)
+                throw MakeStringException(ECLWATCH_INVALID_SEC_MANAGER, MSG_SEC_MANAGER_IS_NULL);
+        }
+        
+        //Verify request
+
+        if (isEmptyString(req.getEnvName()))
+        {
+            throw MakeStringException(-1, "Environment name must be specified");
+            if (strchr(req.getEnvName(), ' '))
+                throw MakeStringException(-1, "Environment name cannot contain spaces or special characters");
+        }
+        if (isEmptyString(req.getEnvOwnerName()))
+            throw MakeStringException(-1, "Environment Owner name must be specified");
+        if (isEmptyString(req.getEnvDescription()))
+            throw MakeStringException(-1, "Environment Description must be specified");
+
+        if (req.getFilesMode()==2  && isEmptyString(req.getCustomFilesBaseDN()))
+            throw MakeStringException(-1, "CustomFilesBaseDN must be specified (ex. 'ou=files,ou=hpcc,dc=myLDAPDomain,dc=com')");
+        if (!req.getGroupsMode()==2 && isEmptyString(req.getCustomGroupsBaseDN()))
+            throw MakeStringException(-1, "CustomGroupsBaseDN must be specified (ex. 'ou=groups,ou=hpcc,dc=myLDAPDomain,dc=com')");
+        if (!req.getUsersMode()==2 && isEmptyString(req.getCustomUsersBaseDN()))
+            throw MakeStringException(-1, "CustomUsersBaseDN must be specified (ex. 'ou=users,ou=hpcc,dc=myLDAPDomain,dc=com')");
+        if (!req.getResourcesMode()==2 && isEmptyString(req.getCustomResourcesBaseDN()))
+            throw MakeStringException(-1, "CustomResourcesBaseDN must be specified (ex. 'ou=smc,ou=espservices,ou=hpcc,dc=myLDAPDomain,dc=com')");
+        if (!req.getWorkunitsMode()==2 && isEmptyString(req.getCustomWorkunitsBaseDN()))
+            throw MakeStringException(-1, "CustomWorkunitsBaseDN must be specified (ex. 'ou=workunits,ou=hpcc,dc=myLDAPDomain,dc=com')");
+
+        // Create OU names
+
+        StringBuffer respFilesBaseDN, respGroupsBaseDN, respUsersBaseDN, respResourcesBaseDN, respWorkunitsBaseDN;
+        formatOUname(respFilesBaseDN, req.getEnvName(), req.getFilesMode(), sharedFilesBaseDN.str(), req.getCustomFilesBaseDN(), "ou=files");
+        formatOUname(respGroupsBaseDN, req.getEnvName(), req.getGroupsMode(), sharedGroupsBaseDN.str(), req.getCustomGroupsBaseDN(), "ou=groups");
+        formatOUname(respUsersBaseDN, req.getEnvName(), req.getUsersMode(), sharedUsersBaseDN.str(), req.getCustomUsersBaseDN(), "ou=users");
+        formatOUname(respResourcesBaseDN, req.getEnvName(), req.getResourcesMode(), sharedResourcesBaseDN.str(), req.getCustomResourcesBaseDN(), "ou=smc,ou=espservices");
+        formatOUname(respWorkunitsBaseDN, req.getEnvName(), req.getWorkunitsMode(), sharedWorkunitsBaseDN.str(), req.getCustomWorkunitsBaseDN(), "ou=workunits");
+
+        // Create HPCC Admin Username
+
+        VStringBuffer respAdminUser("admin_%s", req.getEnvName());
```

Comment:
Is admin_<env> user created?  or is that a TBD?

When the env admin is created I think it should be automatically be given rights to manager private/stand alone ous, but not global/shared ones.. operations should have to grant that.

Reply:
The HPCCAdmins group, and the HPCCAdmin user are created by ESP when it starts, and that user is then granted SMC Access (and nothing more). With SMC Access, they can then add/modify the powers that user/group has.  If we want to change that, I think a separate Jira will be needed,, let me know

Reply:
ok.